### ISSUE TITLE
linter: a disable-next-line naming react/exhaustive-deps or react/rules-of-hooks suppresses every React Compiler diagnostic in the enclosing component

### ISSUE THREAD

**Issue by mikaelh** [oxc-project/oxc#26277]
### What version of Oxlint are you using?

1.80.0 and 1.81.0 (both affected)

### What command did you run?

`oxlint -c .oxlintrc.json case.tsx`

### What does your `.oxlintrc.json` config file look like?

```json
{
  "plugins": ["react"],
  "categories": { "correctness": "error" }
}
```

### What happened?

A `disable-next-line` comment naming **`react/exhaustive-deps`** or **`react/rules-of-hooks`** suppresses **every React Compiler diagnostic in the enclosing component**, not just the following line.

`case.tsx` — two components, each with one `react(purity)` violation. The suppression sits in `B` and names a rule `B` does not violate, with a filler statement so the "next line" is never the violating line:

```tsx
export function A() {
  const a = Date.now();          // react(purity)
  return <div>{a}</div>;
}
export function B() {
  // oxlint-disable-next-line react/exhaustive-deps
  const filler = 0;
  const b = Math.random() + filler;   // react(purity)
  return <div>{b}</div>;
}
```

Expected 2 diagnostics. Got 1 — `B`'s `react(purity)` is gone.

Varying only the rule named in the comment:

| suppression inside `B` | `react()` diagnostics (expected 2) |
|---|---|
| *(none)* | 2 |
| `// eslint-disable-next-line react-hooks/exhaustive-deps` | **1** |
| `// oxlint-disable-next-line react/exhaustive-deps` | **1** |
| `// oxlint-disable-next-line react/rules-of-hooks` | **1** |
| `// oxlint-disable-next-line react/purity` | 2 |
| `// oxlint-disable-next-line react/set-state-in-effect` | 2 |
| `// oxlint-disable-next-line no-unused-vars` | 2 |

Three things narrow it down:

1. **The rule name triggers it, not the comment prefix** — `eslint-` and `oxlint-` behave identically, and the `react-hooks/` prefix behaves the same as `react/`.
2. **Only the enclosing component is affected.** `A`'s diagnostic still reports.
3. **Only React-Compiler-family diagnostics are swallowed.** Adding a `debugger;` statement to `B` alongside its `react(purity)` violation, with the same suppression present, still reports `eslint(no-debugger)` at that line — while `B`'s `react(purity)` stays suppressed.

The whole repro is the 10-line file above plus the 4-line config — nothing project-specific.

### What did you expect to happen?

The comment suppresses the named rule on the following line only, as a `disable-next-line` does for every other rule. `react/purity`, `react/set-state-in-effect` and `no-unused-vars` already behave that way — it is specifically `exhaustive-deps` and `rules-of-hooks` that do not.

### Why it matters

This is silent: the gate stays green and there is no "unused disable directive" warning to hint at it. In our codebase 34 such comments — most of them the ordinary `// eslint-disable-next-line react-hooks/exhaustive-deps` above a deliberately-narrow dep array — were hiding 28 unrelated `react()` findings (`set-state-in-effect`, `refs`, `immutability`, `purity`) across 19 files. The suppressions were legitimate for the rule they named; nobody could have known they were also switching off the rest of the family for the whole component.

The workaround, for anyone who lands here: an `.oxlintrc.json` `overrides` entry setting `"react/exhaustive-deps": "off"` for the file does **not** leak — it suppresses only that rule, and the other `react()` diagnostics in the file are still reported. A comment is the only form that over-suppresses; note that a block `oxlint-disable` / `oxlint-enable` pair is worse still, taking out the whole file.

### Possibly related

#18328 — `react-hooks/exhaustive-deps` anchoring its diagnostic on a different line than eslint. Both look like they could come from where this rule's diagnostics are anchored, though the observable symptoms differ.

**Comment by cbramall** [comment]
Still present in **1.82.0** (latest published at time of writing), so this survives the 1.80/1.81 range in the report.

Ran your `case.tsx` and config unchanged:

```
$ npx oxlint@1.82.0 -c .oxlintrc.json case.tsx
# with the suppression in B
case.tsx:2:13: error react(purity): Cannot call impure function during render

# with the suppression deleted
case.tsx:2:13: error react(purity): Cannot call impure function during render
case.tsx:7:13: error react(purity): Cannot call impure function during render
```

Same result you describe: `A` reports either way, `B`'s diagnostic is swallowed.

One extra data point on impact, from hitting this in a real codebase rather than a reduction. Five `// eslint-disable-next-line react-hooks/exhaustive-deps` comments were hiding **seven unrelated `correctness` errors** — six `react(refs)` and one `react(set-state-in-effect)` — and had been doing so undetected across a long series of code reviews. Component scope is narrower than a file, but it is still wide enough that a suppression at the bottom of a hook hides violations at the top of it, which is where the surprise lives: the comment reads as covering one line and covers everything the component does.

Worth noting the sibling rules behave correctly, which makes the bug harder to spot by analogy — `react/refs` and `react/set-state-in-effect` each suppress exactly the line their directive sits above, verified with a violation elsewhere in the same component that still reported.
